rust/src/tx_builder.rs

```diff
@@ -0,0 +1,164 @@
+        self.input_types.vkeys.insert(hash.clone());
+    }
+    pub fn add_script_input(&mut self, hash: &ScriptHash, input: &TransactionInput) {
+        self.inputs.add(input);
+        self.input_types.scripts.insert(hash.clone());
+    }
+    pub fn add_bootstrap_input(&mut self, hash: &str /* full legacy address */, input: &TransactionInput) {
+        self.inputs.add(input);
+        self.input_types.bootstraps.insert(hash.to_string());
+    }
+
+    pub fn add_output(&mut self, output: &TransactionOutput) {
+        self.outputs.add(output)
+    }
+
+    pub fn set_fee(&mut self, fee: &Coin) {
+        self.fee = Some(fee.clone())
+    }
+
+    pub fn set_certs(&mut self, certs: &Certificates) {
+        self.certs = Some(certs.clone());
+        for cert in &certs.0 {
+            witness_keys_for_cert(cert, &mut self.input_types.vkeys);
+        };
+    }
+
+    pub fn set_withdrawals(&mut self, withdrawals: &Withdrawals) {
+        self.withdrawals = Some(withdrawals.clone());
+        for (withdrawal, _coin) in &withdrawals.0 {
+            self.input_types.vkeys.insert(withdrawal.payment_cred().to_keyhash().unwrap().clone());
+        };
+    }
+
+    pub fn set_metadata(&mut self, metadata: &TransactionMetadata) {
+        self.metadata = Some(metadata.clone())
+    }
+
+    pub fn new() -> Self {
+        Self {
+            inputs: TransactionInputs::new(),
+            outputs: TransactionOutputs::new(),
+            fee: None,
+            ttl: None,
+            certs: None,
+            withdrawals: None,
+            metadata: None,
+            input_types: MockWitnessSet {
+                vkeys: BTreeSet::new(),
+                scripts: BTreeSet::new(),
+                bootstraps: BTreeSet::new(),
+            },
+        }
+    }
+
+    pub fn build(&self) -> Result<TransactionBody, JsValue> {
+        let fee = self.fee.ok_or_else(|| JsValue::from_str("Fee not specified"))?;
+        let ttl = self.ttl.ok_or_else(|| JsValue::from_str("ttl not specified"))?;
+        Ok(TransactionBody {
+            inputs: self.inputs.clone(),
+            outputs: self.outputs.clone(),
+            fee: fee,
+            ttl: ttl,
+            certs: None,
+            withdrawals: None,
+            metadata_hash: match &self.metadata {
+                None => None,
+                Some(x) => Some(utils::hash_metadata(x)),
+            },
+        })
+    }
+
+    pub fn calc_fee(&mut self, linear_fee: &fees::LinearFee) -> Result<Coin, JsValue> {
+        let old_fee = self.fee; // save old value for fee
+
+        match self.fee {
+            // if user explicitly specified a fee already, use that one
+            Some(_x) => {},
+            // otherwise, use the maximum fee possible
+            None => {
+                self.set_fee(&Coin::new(0x1_00_00_00_00));
+            },
+        };
+        let body = self.build()?;
+        let witness_set = TransactionWitnessSet {
+            vkeys: Some(Vkeywitnesses::new()),
```

Comment:
TODO: I need to make these actually be populated with fake witnesses. Need to spend time looking up how to generate fake Byron witnesses, etc.